### DISCUSSION TITLE
Error when compiling on windows

### DISCUSSION THREAD

**Discussion by PabloKench** [randombit/botan#4379] [Q&A] [answered]
I'm starting to use botan in my project, I need the lib on both x64 and x86 but I'm getting an error on Windows to compile

In x64 mode I get zlib error, I am trying to compile as follows

`python configure.py --cc=msvc --os=windows --cpu=x64 --disable-shared --build-targets=static --without-debug-info --prefix=build\x64 --with-zlib`

**Error:**

```
src/lib/compression/zlib/zlib.cpp(14): fatal error C1083: Cannot open include file: 'zlib.h': No such file or directory
NMAKE : fatal error U1077: 'cl   /Zc:preprocessor /std:c++20 /EHs /GR /D_WIN32_WINNT=0x0600 /MD /bigobj /O2 /Oi /Zc:throwingNew /W4 /wd4251 /wd4275 /wd5072 -DBOTAN_IS_BEING_BUILT  /I build/include/public /I build/include/internal  /external:W0 /external:I build/include/external /nologo /c src/lib/compression/zlib/zlib.cpp /Fobuild/obj/lib/compression_zlib.obj' : return code '0x2'
Stop.
```

And in x86 mode 

`python configure.py --cc=msvc --os=windows --disable-cc-tests --cpu=x86 --disable-shared --build-targets=static --without-debug-info --prefix=build\x86 --with-zlib`

When I compile I get the error

```
build/include/public\botan/build.h(503): fatal error C1189: #error:  "Trying to compile Botan configured as x86_32 with non-x86_32 compiler."
NMAKE : fatal error U1077: 'cl   /Zc:preprocessor /std:c++20 /EHs /GR /D_WIN32_WINNT=0x0600 /MD /bigobj /O2 /Oi /Zc:throwingNew /W4 /wd4251 /wd4275 /wd5072 -DBOTAN_IS_BEING_BUILT  /I build/include/public /I build/include/internal  /external:W0 /external:I build/include/external /nologo /c src/lib/asn1/alg_id.cpp /Fobuild/obj/lib/asn1_alg_id.obj' : return code '0x2'
Stop.
```

And a question I wanted to ask is it possible to generate an **sln** file for the Windows compiler?

**Comment by reneme** [comment, marked as answer]
Botan's build system isn't able to find zlib on your system without some extra help. On unix based systems it would simply look at the system's default locations, but this concept doesn't exist in the same way on Windows, unfortunately. Please try providing the include and library locations like this:

```bash
./configure.py [... any other flags ...] --with-zlib --with-external-includedir=C:/your/path/to/zlib/include --with-external-libdir=C:/you/path/to/zlib/lib
```

Regarding the x86 issue, this is a bit a shot in the dark, but my guess: Did you run this in an x86_64 Visual Studio environment? Because, the build system picked up a compiler that doesn't seem to support building for x86_32.
Typically you don't really have to provide `--cc=`, `--os=` and `--cpu=`. It'll just deduce defaults from your environment in most cases.

> is it possible to generate an sln file for the Windows compiler?

Unfortunately not, our build system is fairly low-level and can only generate `Makefile` or `build.ninja`. I would suggest to try using a package manager to provision the dependencies of your project. Good choices for C++ might be Microsoft's "vcpkg" or the independent "conan" package manager. Both come with mature recipes to build Botan on Windows and might get you started more easily.

**Comment by PabloKench** [comment]
I managed to compile it, thanks for the help
